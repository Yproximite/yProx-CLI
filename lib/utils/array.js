/**
 * Should be used in Array#reduce().
 * @param {String} key
 * @return {Function}
 */
module.exports.groupBy = (key) => {
  return (acc, item) => {
    const value = item[key];

    if (!acc[value]) {
      acc[value] = [];
    }

    acc[value].push(item);

    return acc;
  };
};

/**
 * Should be used in Array#reduce().
 * @return {Function}
 */
module.exports.flatten = () => {
  return (acc, values) => {
    acc = acc.concat(values);
    return acc;
  };
};
