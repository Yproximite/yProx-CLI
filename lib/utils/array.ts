/**
 * Should be used in Array#reduce().
 */
export function groupBy(key: string): (acc: { [k: string]: any }, item: any) => { [k: string]: any } {
  return (acc: { [k: string]: any }, item: any) => {
    const value = item[key];

    if (!acc[value]) {
      acc[value] = [];
    }

    acc[value].push(item);

    return acc;
  };
}

/**
 * Should be used in Array#reduce().
 */
export function flatten() {
  return (acc: [], values: []) => {
    // @ts-ignore
    acc = acc.concat(values);
    return acc;
  };
}
