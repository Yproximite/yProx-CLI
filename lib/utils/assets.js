const readEntryFile = (api, entryName, entries) => {
  if (typeof entries === 'string') {
    entries = require(api.resolve(entries));
  }
  if (typeof entries === 'function') {
    entries = entries(api, api.projectOptions);
  }

  return entries.map(entry => {
      entry._name = entryName;
      entry.dest = api.resolve(entry.dest);

      return entry;
    });
};

module.exports.readAssets = (api, args) => {
  let assets = [];

  Object
    .entries(api.projectOptions.assets)
    .forEach(([entryName, entryFile]) => {
      assets = assets.concat(readEntryFile(api, entryName, entryFile));
    });

  Object
    .entries(args)
    .filter(([argKey, argValue]) => argKey.startsWith('filter:'))
    .forEach(([argKey, argValue]) => {
      const attrToFilter = argKey.split(':')[1];

      if (Array.isArray(argValue)) {
        api.logger.log(`Filtering assets where \`['${argValue.join("', '")}'].includes(asset.${attrToFilter})\``);
      } else {
        api.logger.log(`Filtering assets where \`asset.${attrToFilter} === '${argValue}'\``);
      }

      assets = assets.filter(asset => {
        const valueToFilter = asset[attrToFilter];

        if (Array.isArray(argValue)) {
          return argValue.includes(valueToFilter);
        }

        return valueToFilter === argValue;
      });
    });

  return assets;
};
