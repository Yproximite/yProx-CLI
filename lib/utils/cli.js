const chalk = require('chalk');
const CLI_BIN = 'yprox-cli';

module.exports.displayHelp = (commands) => {
  console.log(chalk`\n  Usage: {green ${CLI_BIN} [command] [options]}\n`);

  displayOptions({
    '--help': 'display this help',
    '--version': 'display version',
  });

  displayCommands(commands);

  console.log(chalk`  run {green ${CLI_BIN} [command] --help} for usage of a specific command\n`);
};

module.exports.displayCommandHelp = (commandName, command) => {
  console.log(chalk`\n  Usage: {green ${CLI_BIN} ${commandName} [options]}\n`);
  displayOptions(command.opts.options);
};

function displayCommands(commands) {
  displaySectionTable('Commands', commands, (entry) => [entry[0], entry[1].opts.description]);
}

function displayOptions(options) {
  displaySectionTable('Options', options, (entry) => [entry[0], entry[1]]);
}

function displaySectionTable(title, dataObject, handleEntry) {
  if (!dataObject || Object.keys(dataObject).length === 0) {
    return;
  }

  const maxLen = Math.max(...Object.keys(dataObject).map(key => key.length));

  console.log(`  ${title}:\n`);

  Object.entries(dataObject).forEach(entry => {
    const [name, value] = handleEntry(entry);
    console.log(chalk`    {blue ${name.padEnd(maxLen)}} ${value}`);
  });

  console.log();
}
