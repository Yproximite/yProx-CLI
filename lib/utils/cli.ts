import chalk from 'chalk';

const CLI_BIN = 'yprox-cli';

function displaySectionTable(title: string, dataObject: { [k: string]: any }, handleOption: (opts: any[]) => string[]): void {
  if (!dataObject || Object.keys(dataObject).length === 0) {
    return;
  }

  const maxLen = Math.max(...Object.keys(dataObject).map(key => key.length));

  console.log(`  ${title}:\n`);

  Object.entries(dataObject).forEach(option => {
    const [name, value] = handleOption(option);
    console.log(chalk`    {blue ${name.padEnd(maxLen)}} ${value}`);
  });

  console.log();
}

function displayCommands(commands: CLICommands): void {
  displaySectionTable('Commands', commands, opts => [opts[0], opts[1].opts.description]);
}

function displayOptions(options: CLICommandOptions): void {
  displaySectionTable('Options', options, opts => [opts[0], opts[1]]);
}

export function displayHelp(commands: CLICommands): void {
  console.log(chalk`\n  Usage: {green ${CLI_BIN} [command] [options]}\n`);

  displayOptions({
    '--help': 'display this help',
    '--version': 'display version',
  });

  displayCommands(commands);

  console.log(chalk`  run {green ${CLI_BIN} [command] --help} for usage of a specific command\n`);
}

export function displayCommandHelp(commandName: string, command: CLICommand): void {
  console.log(chalk`\n  Usage: {green ${CLI_BIN} ${commandName} [options]}\n`);
  displayOptions(command.opts.options);
}
