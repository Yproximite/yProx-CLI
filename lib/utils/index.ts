const watcher = require('./watcher');

module.exports = {
  watcher,
};
