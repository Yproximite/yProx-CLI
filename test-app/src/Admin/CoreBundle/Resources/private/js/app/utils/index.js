import apiInterface from './apiInterface';
import Routing from './routing';
import provideComponents from './provideComponents';
import provideDirectives from './provideDirectives';
import provideFilters from './provideFilters';
import escapeRegex from './escapeRegex';
import deepClone from './deepClone';

export default {
  apiInterface,
  Routing,
  provideComponents,
  provideDirectives,
  provideFilters,
  escapeRegex,
  deepClone,
};
