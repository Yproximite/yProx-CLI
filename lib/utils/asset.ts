import { Asset } from '../../types';
import { Entry } from '../../types/entry';
import API from '../API';
import { ensureArray } from './array';

export function readAssetDef(api: API, assetName: string, asset: Asset): Entry[] {
  let entries = asset;

  if (typeof entries === 'string') {
    entries = require(api.resolve(entries));
  }
  if (typeof entries === 'function') {
    entries = entries(api, api.projectOptions);
  }
  entries = entries as Entry[];

  return entries.map(entry => {
    return {
      ...entry,
      _name: assetName,
      src: ensureArray(entry.src).map(src => api.resolve(src)),
      dest: api.resolve(entry.dest),
      sourceMaps: api.isProduction(),
    };
  });
}
