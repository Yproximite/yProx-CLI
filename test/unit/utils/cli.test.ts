import { displayCommandHelp, displayHelp } from '../../../lib/utils/cli';

describe('utils: cli', () => {
  const commands = {
    build: {
      name: 'build',
      fn: () => Promise.resolve(1),
      opts: {
        description: 'Build files',
        usage: 'yprox-cli build [options]',
        options: { '--lint': 'Lint before build' },
      },
    },
    lint: {
      name: 'lint',
      fn: () => Promise.resolve(2),
      opts: {
        description: 'Lint files',
        usage: 'yprox-cli lint [options]',
        options: { '--fix': 'Fix some errors' },
      },
    },
  };

  describe('show global help', () => {
    it('should show global help', () => {
      let output = '';
      console.log = jest.fn(data => (output += `${data || ''}\n`));

      displayHelp(commands);
      expect(output).toMatchSnapshot();
    });
  });

  describe('show a specific command help', () => {
    it('should show help for `build` command', () => {
      let output = '';
      console.log = jest.fn(data => (output += `${data || ''}\n`));

      displayCommandHelp('build', commands.build);
      expect(output).toMatchSnapshot();
    });

    it('should show help for `lint` command', () => {
      let output = '';
      console.log = jest.fn(data => (output += `${data || ''}\n`));

      displayCommandHelp('lint', commands.lint);
      expect(output).toMatchSnapshot();
    });
  });
});
