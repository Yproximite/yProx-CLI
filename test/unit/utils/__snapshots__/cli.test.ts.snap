// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils: cli show a specific command help should show help for \`build\` command 1`] = `
"
  Usage: yprox-cli build [options]

  Options:

    --lint Lint before build

"
`;

exports[`utils: cli show a specific command help should show help for \`lint\` command 1`] = `
"
  Usage: yprox-cli lint [options]

  Options:

    --fix Fix some errors

"
`;

exports[`utils: cli show global help should show global help 1`] = `
"
  Usage: yprox-cli [command] [options]

  Options:

    --help    display this help
    --version display version

  Commands:

    build Build files
    lint  Lint files

  run yprox-cli [command] --help for usage of a specific command

"
`;
